### ISSUE TITLE
Spell check body in typescriptreact (*.tsx)

### ISSUE THREAD

**Issue by borekb** [bartosz-antosik/vscode-spellright#117]
I think SpellRight doesn't check text inside "HTML" / JSX elements in `*.tsx` files. This is my config:

```
    "spellright.spellContext": "body comments strings",
    "spellright.spellContextByClass": {
        "markdown": "body",
        "typescript": "comments",
        "json": "comments",
        "html": "body",
        "typescriptreact": "body comments strings",
    },
```

This is the spell check:

![image](https://user-images.githubusercontent.com/101152/36484786-a800aaa0-171a-11e8-8801-2d6bca36a9a0.png)

I'm not sure how SpellRight works but detecting body might be quite tricky in JSX..

**Comment by bartosz-antosik** [comment]
At first glance looks terrible. 

As I know null about typescriptreact - could you please point to some resource describing this syntax plus some "pasteable" example(s)?

**Comment by borekb** [comment]
Yeah.. :) React is a mix of JavaScript and HTML tags, looks weird at first but it's actually surprisingly good in practice. I guess checking out some existing React + TypeScript project would give you a lot of example, for example, this is ours:

https://github.com/versionpress/versionpress/blob/129bf9724fa468cfcdbb86df6f7ddfa86c46889f/frontend/src/components/welcome-panel/WelcomePanel.tsx

```jsx
import * as React from 'react';

import './WelcomePanel.less';

interface WelcomePanelProps {
  onHide(): void;
}

const WelcomePanel: React.StatelessComponent<WelcomePanelProps> = ({ onHide }) => (
  <div className='WelcomePanel welcome-panel'>
    <a
      className='welcome-panel-close'
      href='#'
      onClick={e => { e.preventDefault(); onHide(); }}
    >
      Dismiss
    </a>
    <div className='welcome-panel-content'>
      <h3>Welcome!</h3>
      <p className='about-description'>
        Below is the main VersionPress table which will grow as changes are made to this site.
        You can <strong>Undo</strong> specific changes from the history or <strong>Roll back</strong> the site
        entirely to a previous state.
      </p>
    </div>
  </div>
);

export default WelcomePanel;
```

There are many more components if you need examples. Here is [React's Hello World](https://reactjs.org/docs/hello-world.html).

**Comment by bartosz-antosik** [comment]
Looks really tough. And it certainly breaks the principle of separating presentation from processing!

I will have to look and meditate. As usually.

**Comment by bartosz-antosik** [comment]
And just BTW, `body` is not adequate for every class. Source code documents do not respect `body` at all. So generally speaking not every document is obliged to support all of the classes.

**Comment by borekb** [comment]
So what would be the appropriate config for JSX / React? I guess it's the same for PHP + HTML, simply speaking, I'd like to spell check everything inside HTML tags. Is `body` good for that?

**Comment by bartosz-antosik** [comment]
No, no I meant something different. Just that not every parser is obliged to support all of the parts, e.g. like plaintext parser is not expected to support `strings` or `comments`. `body` would be great for this, esp. if someone would come up with a PR for this...

**Comment by bartosz-antosik** [comment]
Bernard! I did three approaches to the issue over the time. I am sorry, but I am about to give up! Heuristics I have invented about this have flaws and are hard to manage. It would call for a proper parser just for this type of documents (with embedded HTML/XML parts) which Is beyond me at the moment.

I will close the thread and maybe revive it in the future if something happens (e.g. VSCode gives extensions access to syntactical information).
